alts: Fix TsiSocket doRead to obey read limits

yihuazhang opened the pull request:
This PR resolves the second issue described in #15296.

antoniovicente commented on test/extensions/transport_sockets/alts/tsi_socket_test.cc:
```diff
@@ -382,6 +383,152 @@ TEST_F(TsiSocketTest, HandshakeWithInternalError) {
+
+  EXPECT_CALL(*client_.raw_socket_, doWrite(_, false));
+  expectIoResult({Network::PostIoAction::KeepOpen, 21UL, false},
+                 client_.tsi_socket_->doWrite(client_.write_buffer_, false));
```
nit: I think that the number of bytes written should be the length of ClientToServerData.  Same about the number of bytes read.

yihuazhang replied:
Done.

antoniovicente commented on test/extensions/transport_sockets/alts/tsi_socket_test.cc:
```diff
@@ -382,6 +383,152 @@ TEST_F(TsiSocketTest, HandshakeWithInternalError) {
+  expectIoResult({Network::PostIoAction::KeepOpen, 21UL, false},
+                 client_.tsi_socket_->doWrite(client_.write_buffer_, false));
+
+  EXPECT_CALL(*server_.raw_socket_, doRead(_)).WillOnce(Invoke([&](Buffer::Instance& buffer) {
```
I would recommend adding "InSequence s;" before this line to verify that the expectations happen in the expected order.

yihuazhang replied:
SG. Done.

yihuazhang replied:
Done.

antoniovicente commented on test/extensions/transport_sockets/alts/tsi_socket_test.cc:
```diff
@@ -382,6 +383,152 @@ TEST_F(TsiSocketTest, HandshakeWithInternalError) {
+
+  EXPECT_EQ(ClientToServerData, server_.read_buffer_.toString());
+
+  EXPECT_CALL(*server_.raw_socket_, doRead(_))
```
2 reads that return 0 bytes in a row seems unexpected.  The call to shouldDrainReadBuffer also seems unexpected.

yihuazhang replied:
Done.

antoniovicente commented on test/extensions/transport_sockets/alts/tsi_socket_test.cc:
```diff
@@ -382,6 +383,152 @@ TEST_F(TsiSocketTest, HandshakeWithInternalError) {
+}
+
+TEST_F(TsiSocketTest, DoReadError) {
+  initialize(nullptr, nullptr);
```
I see that server_.raw_socket_ is a NiceMock, should we consider removing NiceMock in order to get the StrictMock default in these tests?

https://github.com/envoyproxy/envoy/blob/a7f46e02a0857926aace227fce2056164b5a6f10/test/extensions/transport_sockets/alts/tsi_socket_test.cc#L47

yihuazhang replied:
Done.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -141,6 +143,9 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
 
+    // Reset the watermarks with actual negotiated max frame size.
+    read_buffer_.setWatermarks(
+        std::max<size_t>(actual_frame_size_to_use_, callbacks_->connection().bufferLimit()));
```
actual_frame_size_to_use_ + frame_overhead

yihuazhang replied:
Not needed per previous comment.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,10 +170,60 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
   Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
   if (!end_stream_read_ && !read_error_) {
-    result = raw_buffer_socket_->doRead(raw_read_buffer_);
+    result = raw_buffer_socket_->doRead(read_buffer_);
```
I think that we may need to read in a loop while handling the handshake now that we are setting a lower initial watermark limit on the buffer.  Without it, resumption may not happen correctly for very large handshake messages.

I think that the current implementation of watermark callbacks on the read_buffer_ ends up hiding the large handshake resumption failure as calling readDisable(true) followed by readDisable(false) effectively re-registers the socket for events.

yihuazhang replied:
Or how about we create another read_buffer that does not have any watermark set and use it just for the handshake? I think it will be much simpler.

antoniovicente replied:
The issue is with the transition from handshake to operations past the end of the handshake.  By allowing unbounded reads during the handshake we could end up having to copy a lot of post handshake protected bytes when handling handshake complete back to the raw buffer because those bytes were read into the unbounded buffer used during the handshake.

We need to run in a loop during both the handshake and post-handshake.  Using a buffer without bounds for the handshake doesn't simplify much.

yihuazhang replied:
I do not think "unbounded reads" will happen during the handshake because we are enforcing the read limit on the read_buffer (which will contain all handshake messages and unused protected bytes received from peer) via shouldDrainReadBuffer(). It then passes the data in read_buffer into the raw buffer ([here](https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/alts/tsi_socket.cc#L71)), which although does not have any read limit, thanks to the limit set on the read_buffer, the raw buffer will not hold unlimitedly large data.

antoniovicente reviewed (commented):
Thanks again for taking on these improvements to the TsiSocket implementation.  I think I ran into a few additional edge cases while looking at the code more closely as part of this PR.  Let me know what you think about the tsi_result case and if there's ways in which I can help.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -17,16 +17,18 @@ TsiSocket::TsiSocket(HandshakerFactory handshaker_factory, HandshakeValidator ha
     : TsiSocket(handshaker_factory, handshake_validator,
-                std::make_unique<Network::RawBufferSocket>()) {}
+                std::make_unique<Network::RawBufferSocket>()) {
+  read_buffer_.setWatermarks(default_max_frame_size_);
```
By "frame overhead", do you mean tag length? In gRPC ALTS zero-copy frame protector, it computes the max number of bytes allowed to protect for each frame ([link](https://github.com/grpc/grpc/blob/master/src/core/tsi/alts/zero_copy_frame_protector/alts_grpc_record_protocol_common.cc#L167)) by subtracting the tag length and other overhead from the max frame size ([link](and it is computed as subtracting the tag length and frame header)). It means that if you receive a full frame, you should be able to decrypt it.

antoniovicente replied:
Good to know.  Ok.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -68,7 +70,7 @@ void TsiSocket::doHandshakeNext() {
   handshaker_next_calling_ = true;
   Buffer::OwnedImpl handshaker_buffer;
-  handshaker_buffer.move(raw_read_buffer_);
+  handshaker_buffer.move(read_buffer_);
```
Ack.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -43,15 +45,15 @@ absl::string_view TsiSocket::failureReason() const {
   ASSERT(!handshake_complete_);
   ENVOY_CONN_LOG(debug, "TSI: doHandshake", callbacks_->connection());
-  if (!handshaker_next_calling_ && raw_read_buffer_.length() > 0) {
+  if (!handshaker_next_calling_ && read_buffer_.length() > 0) {
```
Done. Will be addressed in the next commit.

antoniovicente replied:
What do you mean by next commit?  I think you may need to push your local changes so we can get them reviewed as part of this PR.

yihuazhang replied:
Sorry for the confusion. By "next commit", I mean the change that addresses all your comments this round, and will be pushed for a review. It is for sure a part of this PR.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -124,7 +126,7 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
     ASSERT(status == TSI_OK);
     if (unused_byte_size > 0) {
-      raw_read_buffer_.prepend(
+      read_buffer_.prepend(
```
Done. Will be addressed in the next commit.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,10 +170,60 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+
+void TsiSocket::onReadBufferHighWatermark() {
+  if (callbacks_->connection().state() == Network::Connection::State::Open) {
+    callbacks_->connection().readDisable(true);
```
Ack. Done.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,10 +170,60 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+        end_stream_read_ = result.end_stream_read_;
+        read_error_ = result.action_ == Network::PostIoAction::Close;
+        // No data is read or read error happens.
+        if (result.bytes_processed_ == 0) {
```
I think a more serious problem is a connection can get closed ([link](https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/alts/tsi_socket.cc#L175-L176)) without unprotecting data in read_buffer_ and the correct way is to skip this part ([link](https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/alts/tsi_socket.cc#L170-L185)) when handshake is complete, and only rely on `DoRead()` implemented in `repeatReadAndUnprotect`.

yihuazhang replied:
Also, I think we still need `if (result.bytes_processed_ == 0) {` branch to exit the loop early.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,10 +170,60 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+      if (!end_stream_read_ && !read_error_) {
+        result = raw_buffer_socket_->doRead(read_buffer_);
+        end_stream_read_ = result.end_stream_read_;
+        read_error_ = result.action_ == Network::PostIoAction::Close;
```
I do not think we should exit because we may have successfully read some data before an error happens, and we need to unprotect them.

antoniovicente replied:
I can see it being useful to unprotect.  We should break after unprotect if result.action_ == Network::PostIoAction::Close

yihuazhang replied:
Done.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,10 +170,60 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+          ENVOY_CONN_LOG(debug, "TSI: unprotected buffer left: {} result: {}",
+                         callbacks_->connection(), read_buffer_.length(),
+                         tsi_result_to_string(status));
+          total_bytes_processed += read_size - read_buffer_.length();
```
I see. Thanks for the clarification. It will be addressed in the next commit.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -194,13 +249,15 @@ Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
-                   raw_read_buffer_.length());
-    tsi_result status = frame_protector_->unprotect(raw_read_buffer_, buffer);
+                   read_buffer_.length());
+    tsi_result status = frame_protector_->unprotect(read_buffer_, buffer);
```
Done.

yihuazhang commented on test/extensions/transport_sockets/alts/tsi_socket_test.cc:
```diff
@@ -382,6 +383,152 @@ TEST_F(TsiSocketTest, HandshakeWithInternalError) {
+        return result;
+      }))
+      .WillOnce(Invoke([&](Buffer::Instance& buffer) {
+        Network::IoResult result = {Network::PostIoAction::Close, 0UL, false};
```
Done.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -194,13 +249,15 @@ Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
+    uint64_t read_size = read_buffer_.length();
     ENVOY_CONN_LOG(debug, "TSI: unprotecting buffer size: {}", callbacks_->connection(),
-                   raw_read_buffer_.length());
-    tsi_result status = frame_protector_->unprotect(raw_read_buffer_, buffer);
```
Done.

yihuazhang commented:
Thanks much @antoniovicente fo detailed review! I think I addressed all of your comments except for the one w.r.t. calling doRead() in a loop for handshake message. PLMK what you think.

yihuazhang commented:
Friendly ping @antoniovicente

antoniovicente reviewed (commented):
Sorry for the delay.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+  uint64_t total_bytes_processed = 0;
+
+  do {
+    // Do unprotect.
```
Good point. I will skip the unprotect if raw_read_buffer_ is empty.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+    uint64_t prev_size = buffer.length();
+    ENVOY_CONN_LOG(debug, "TSI: unprotecting buffer size: {}", callbacks_->connection(),
+                   raw_read_buffer_.length());
+    tsi_result status = frame_protector_->unprotect(raw_read_buffer_, buffer);
```
Done.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+                   raw_read_buffer_.length());
+    tsi_result status = frame_protector_->unprotect(raw_read_buffer_, buffer);
+    if (status != TSI_OK) {
+      break;
```
+1. Although the API contract for  [`tsi_zero_copy_grpc_protector_unprotect`](https://github.com/grpc/grpc/blob/master/src/core/tsi/transport_security_grpc.h#L46-L54) does not specify what to do if result is not OK, by looking at the types of error the API may return (e.g., data corruption, decryption failure), it is preferable to make it consistent with ssl_socket.cc which [sets the action_ to Network::PostIoAction::Close](https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/tls/ssl_socket.cc#L149-L152). Done.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+      break;
+    } else {
+      // Do another read.
+      if (!end_stream_read_ && !read_error_) {
```
Done.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+    } else {
+      // Do another read.
+      if (!end_stream_read_ && !read_error_) {
+        result = raw_buffer_socket_->doRead(raw_read_buffer_);
```
Done.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+        break;
+      }
+    }
+  } while (true);
```
Done.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
 Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
   Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
-  if (!end_stream_read_ && !read_error_) {
+  if (!handshake_complete_ && !end_stream_read_ && !read_error_) {
```
Agree on redundant !handshake_complete_ check. But I do not think we should remove other checks for !handshake_complete_ out of the body of "if (!handshake_complete_)" because they can be used to return early without doing handshake.

yihuazhang commented on source/extensions/transport_sockets/alts/tsi_socket.h:
```diff
@@ -90,10 +97,10 @@ class TsiSocket : public Network::TransportSocket,
   Network::TransportSocketPtr raw_buffer_socket_;
 
-  Envoy::Buffer::OwnedImpl raw_read_buffer_;
+  Buffer::WatermarkBuffer raw_read_buffer_{[]() -> void {}, []() -> void {}, []() -> void {}};
```
Done.

yihuazhang commented on test/extensions/transport_sockets/alts/tsi_socket_test.cc:
```diff
@@ -44,13 +46,13 @@ class TsiSocketTest : public testing::Test {
 
   void initialize(HandshakeValidator server_validator, HandshakeValidator client_validator) {
-    server_.raw_socket_ = new NiceMock<Network::MockTransportSocket>();
+    server_.raw_socket_ = new StrictMock<Network::MockTransportSocket>();
```
It gives me an "incompatible pointer types assignment" error if omitting "Strictmock<>" :(

antoniovicente replied:
You also need to remove StrictMock from line 182.

yihuazhang commented:
Thanks @antoniovicente for the comments. They are all addressed. PTAL.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -197,47 +202,54 @@ Network::IoResult TsiSocket::repeatReadAndUnprotect(Buffer::Instance& buffer,
+    // Do another read.
+    result = raw_buffer_socket_->doRead(raw_read_buffer_);
+    end_stream_read_ = result.end_stream_read_;
+    read_error_ = result.action_ == Network::PostIoAction::Close;
```
Replace with: readFromRawSocket()

yihuazhang replied:
Opps, my bad. Done.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -193,14 +258,7 @@ Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
-    ENVOY_CONN_LOG(debug, "TSI: unprotected buffer left: {} result: {}", callbacks_->connection(),
-                   raw_read_buffer_.length(), tsi_result_to_string(status));
-    result.bytes_processed_ = read_size - raw_read_buffer_.length();
+    return repeatReadAndUnprotect(buffer, result);
```
The ENVOY_CONN_LOG on line 264 is no longer hit due to early returns like this one.

yihuazhang replied:
Good point. I moved it inside `repeatReadAndUnprotect`.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,9 +172,55 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+    ENVOY_CONN_LOG(debug, "TSI: unprotecting buffer size: {}", callbacks_->connection(),
+                   raw_read_buffer_.length());
+    tsi_result status = frame_protector_->unprotect(raw_read_buffer_, buffer);
+    if (status != TSI_OK) {
```
Consider adding some debug logging when status != TSI_OK

yihuazhang replied:
Done.

antoniovicente reviewed (dismissed):
Thanks for the fixes.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,26 +172,86 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+        ENVOY_CONN_LOG(debug, "TSI: unprotect failed: status: {}", callbacks_->connection(),
+                       status);
+        result.action_ = Network::PostIoAction::Close;
+        break;
```
No test coverage for this branch, could you add tests?

https://storage.googleapis.com/envoy-pr/0357f2f/coverage/source/extensions/transport_sockets/alts/tsi_socket.cc.gcov.html

yihuazhang replied:
Done.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,26 +172,86 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+    // Read error happens in the previous read.
+    if (read_error_) {
+      result.action_ = Network::PostIoAction::Close;
+      break;
```
No test coverage for this branch, could you add tests?

yihuazhang replied:
It turns out that this if branch is redundant due to `if (result.action_ == Network::PostIoAction::Close)` brach that we have earlier. Because `read_error_` returning true indicates that `result.action_` is always   `Network::PostIoAction::Close`.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -193,14 +260,7 @@ Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
+    return repeatReadAndUnprotect(buffer, result);
   }
 
   ENVOY_CONN_LOG(debug, "TSI: do read result action {} bytes {} end_stream {}",
```
No longer covered by tests.

https://storage.googleapis.com/envoy-pr/0357f2f/coverage/source/extensions/transport_sockets/alts/tsi_socket.cc.gcov.html

yihuazhang replied:
We should not reach this code path any more because handshake_complete_ is always true at this point. Code is updated. PTAL.

antoniovicente commented on test/extensions/transport_sockets/alts/tsi_socket_test.cc:
```diff
@@ -14,9 +14,11 @@ namespace TransportSockets {
 using testing::NiceMock;
 using testing::Return;
 using testing::ReturnRef;
+using testing::StrictMock;
```
From clang tidy: Message: using decl 'StrictMock' is unused

yihuazhang replied:
Done.

antoniovicente commented on source/extensions/transport_sockets/alts/tsi_socket.cc:
```diff
@@ -165,26 +172,86 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
+    if (callbacks_->shouldDrainReadBuffer()) {
+      callbacks_->setTransportSocketIsReadable();
+      break;
+    }
```
I think that the rif (callbacks_->shouldDrainReadBuffer()) { checks need to be protected by if (raw_read_buffer_.length() > 0) {

Please move this block inside the raw_read_buffer_ check.

Handling close action outside the raw_read_buffer_ length check seems correct.

yihuazhang replied:
Done.

antoniovicente reviewed (approved):
Looks good.  A few final nits, but I think we could merge this change as is.

yihuazhang commented on test/extensions/transport_sockets/alts/tsi_socket_test.cc:
```diff
@@ -120,6 +119,21 @@ class TsiSocketTest : public testing::Test {
     return frame;
   }
 
+  std::string makeInvalidTsiFrame() {
```
Done.

antoniovicente reviewed (approved):
Thanks.
